artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -410,6 +418,74 @@ pub fn byteslice(
     length: Option<Value>,
 ) -> Result<Value, Error> {
     let s = unsafe { super::String::unbox_from_value(&mut value, interp)? };
+    let maybe_range = if length.is_none() {
+        index.is_range(interp, s.char_len() as i64)?
+    } else {
+        None
+    };
+    match maybe_range {
+        None => {}
+        Some(protect::Range::Out) => return Ok(Value::nil()),
+        Some(protect::Range::Valid { start: index, len }) => {
+            let index = if let Ok(index) = usize::try_from(index) {
+                Some(index)
+            } else {
+                index
+                    .checked_neg()
+                    .and_then(|index| usize::try_from(index).ok())
+                    .and_then(|index| s.len().checked_sub(index))
+            };
+            let index = match index {
+                None => return Ok(Value::nil()),
+                Some(index) if index > s.len() => return Ok(Value::nil()),
```

Comment:
I think my understanding of how mruby unwraps into `Range::Valid` isn't clear. Should mruby take care of these checks and provide a `Range::Out` instead?

Reply:
I think you're probably right let me check this case in the repl ...

```
artichoke 0.1.0-pre.0 (2022-07-23 revision 5844) [x86_64-apple-darwin]
[rustc 1.62.1 (e092d0b6b 2022-07-16) on x86_64-apple-darwin]
>>> s = "Abc"
=> "Abc"
>>> s.byteslice(4..6)
[artichoke-backend/src/extn/core/string/trampoline.rs:440] Value::nil() = Value(
    nil,
)
=> nil
>>> s.byteslice(10..20)
[artichoke-backend/src/extn/core/string/trampoline.rs:440] Value::nil() = Value(
    nil,
)
=> nil
```

Somehow this case is getting hit.

Reply:
After taking a peek at the code, the given length argument is only used for wrapping negative start and end of the `Range`.

As long as these come out postive, regardless of whether they overlap with `0..len`, `MRB_RANGE_OK` is returned.

https://github.com/artichoke/artichoke/blob/7bf3f033e020dd60f4b5fd5974dcb19af2617f6d/artichoke-backend/vendor/mruby/src/range.c#L456-L488